[FLINK-23354][blob] Limit the size of ShuffleDescriptors in PermanentBlobCache on TaskExecutor

Thesharing opened the pull request:
## What is the purpose of the change

*This pull request intends to add a limit size for ShuffleDescriptors stored in PermanentBlobCache on TaskExecutors. If the size limits exceeds, blob caches are deleted in LRU order. See FLINK-23354 for more details.*

*This pull request is the descendant of FLINK-23218.*

## Brief change log

  - *Introduce BlobCacheSizeTracker to track the size of ShuffleDescriptors stored in PermanentBlobCache on TaskExecutors.*
  - *Introduce `PermanentBlobService#readFile` interface. In PermanentBlobCache, `readFile` makes sure that the file is fully read in the same write lock as the file is accessed. This avoids the scenario that the path is returned as the file is deleted concurrently by other threads. Also `readFile` will track the blobs with BlobCacheSizeTracker. If the size exceeds the limit, PermanentBlobCache will first attempt to delete the files in LRU order and then add the new blob.*

## Verifying this change

This change added tests and can be verified as follows:

  - *Added unit tests for BlobCacheSizeTracker.*
  - *Extended `BlobCacheCleanupTest#testPermanentBlobDeferredCleanup` for cleaning up the track when the job is released.*
  - *Added `DefaultExecutionGraphDeploymentWithSizeLimitTest` to make sure deployment works well when the size limit exceeds.*
  - *Manually verified the change by running an end-to-end test. The job has two JobVertices. One is source, while the other is sink. The parallelism is 8k and 16k. They are connected with an all-to-all edge. Both streaming mode and batch mode are tested.*

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (**yes** / no / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/AbstractBlobCache.java:
```diff
@@ -249,6 +278,28 @@ public void close() throws IOException {
+                    "Failed to delete locally cached BLOB {} at {}",
+                    blobKey,
+                    localFile.getAbsolutePath());
+            return false;
```
Why do we have a return value but add no handling for it?

Thesharing replied:
> Why do we have a return value but add no handling for it?

Since the failure of deleting files is tolerable and the log has been outputted here, we can remove the return value.

zhuzhurk commented on flink-core/src/main/java/org/apache/flink/configuration/BlobServerOptions.java:
```diff
@@ -118,4 +118,10 @@
+    public static final ConfigOption<Long> BLOB_CACHE_SIZE_LIMIT =
+            key("blob.cache.size.limit")
+                    .defaultValue(1_024 * 1_024 * 1_024 * 10L)
+                    .withDescription("The maximum size limit of blob cache for each TaskExecutor.");
```
the description needs clearly state how it affects permanent blob and transient blob size.

Thesharing replied:
> the description needs to clearly state how it affects permanent blob and transient blob size limit.

I've updated the descriptions here.

Thesharing replied:
This option has been removed.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -0,0 +1,157 @@
+
+    private final Object lock = new Object();
+
+    private final long capacity;
```
name it as `sizeLimit` would be easier for understanding

Thesharing replied:
> name it as `sizeLimit` would be easier for understanding

Thank you for pointing this out. Resolved.

Thesharing commented:
Hi @zhuzhurk. I've refactor the code based on the lastest idea we've discussed before. Would you mind re-reviewing it if you got any free time? Thank you in advance.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:
```diff
@@ -0,0 +1,286 @@
+
+    private static final Random random = new Random();
+
+    private static final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
Can be simplified to be: 
```
@Rule public final TemporaryFolder temporaryFolder = new TemporaryFolder();
```
And the `setUp()` and `cleanUp ()` will not be needed.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Test
+    public void testSizeLimitTrackerStatusUpdate() {
```
Can be simplified to be `testUpdate `

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Test
+    public void testTrackPermanentBlobCache() throws Exception {
```
It is test of `PermanentBlobCache` and should not be in `BlobCacheSizeTrackerTest`.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:
```diff
@@ -0,0 +1,286 @@
+    }
+
+    @Test
+    public void testGetPermanentCacheWithSizeLimitConcurrently() throws Exception {
```
It is test of `PermanentBlobCache` and should not be in `BlobCacheSizeTrackerTest`.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    @VisibleForTesting
+    Set<BlobKey> getBlobKeysByJobId(JobID jobId) {
```
Seems this method is never used?

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    /** Remove the BLOB from the tracker. */
+    public void untrack(@Nullable JobID jobId, BlobKey blobKey) {
```
It can be private

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -0,0 +1,177 @@
+    }
+
+    /** Unregister all the tracked BLOBs related to given job. */
+    public void unregisterJob(JobID jobId) {
```
I prefer the name to be `untrackAll` for easier understanding and also because there is no `registerJob()`

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -0,0 +1,177 @@
+        checkArgument(size >= 0);
+
+        synchronized (lock) {
+            caches.put(Tuple2.of(jobId, blobKey), size);
```
`put` -> `putIfAbsent` and increate total only if it is newly added

Thesharing reviewed (commented):
Thank you so much for your review, @zhuzhurk. You've provided such great suggestions.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java:
```diff
@@ -45,4 +46,21 @@
+    /**
+     * Directly read the file associated with the provided job ID and blob key.
+     *
+     * <p>Compared to {@code getFile}, {@code readFile} makes sure that the file is fully read when
```
It's better to illustrate here that why we need to add this interface.

Thesharing replied:
> It's better to illustrate here that why we need to add this interface.

More detailed comments are added.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobService.java:
```diff
@@ -50,13 +50,20 @@
+     * <p>Compared to {@code getFile}, {@code readFile} will attempt to read the entire file after
+     * retrieving it. If file reading and file retrieving is done in the same WRITE lock, it can
+     * avoid the scenario that the path to the file is deleted concurrently by other threads at the
+     * same time as the file is retrieved and read.
```
maybe "at the same time as the file is retrieved and read" -> "when the file is retrieved but not read yet"?

Thesharing replied:
Thank you for pointing this out. Resolved.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/executiongraph/DefaultExecutionGraphDeploymentWithSizeLimitTest.java:
```diff
@@ -0,0 +1,185 @@
+    }
+
+    @Test
+    public void testDeployTasksWithMinimumSizeLimit() throws Exception {
```
Ok. Then how about changing the test name to `testDeployMultipleTasksWithBlobCacheSizeLimit`?

Thesharing replied:
Sure. Resolved.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -187,6 +211,139 @@ public File getFile(JobID jobId, PermanentBlobKey key) throws IOException {
+
+        for (Tuple2<JobID, BlobKey> key : blobsToDelete) {
+            deleteFile(key.f0, key.f1);
+            blobCacheSizeTracker.untrack(key);
```
>>> The only risk is that we will try to delete the blob every time a new blob inserts until the deletion is successful

I think it's better to have to expose the problem. And it also enables retry of failed deletions.

Thesharing replied:
I'm not sure the meaning of exposing the problem. If the deletion is failed, `deleteFile` will output a warning log.

zhuzhurk replied:
I mean I think it's not "risk is that we will try to delete the blob every time a new blob inserts until the deletion is successful". It's good to log warnings and retry deletions in this case.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -187,6 +211,139 @@ public File getFile(JobID jobId, PermanentBlobKey key) throws IOException {
+
+    private void checkLimitAndMoveFile(
+            File incomingFile,
+            @Nullable JobID jobId,
```
We should also add `checkNotNull` for it.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
+                            size);
+                }
+            } else {
                 LOG.warn(
```
Is it possible to happen frequently. If yes, I prefer to make it a debug log.

Thesharing replied:
I think this warning shouldn't happen frequently. In `BlobUtil.moveTempFileToStore`, if the blob key duplicates, a warning will be outputted, too. Furthermore, the previous warning ("The overall size of ShuffleDescriptors in PermanentBlobCache exceeds ...") should be warning, too. The overall size exceeds if and only if the size of blob exceeds 100MiB. In this scenario the user should be warned about it.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
+                                    + "the limit. Limit = [{}], Current: [{}], "
+                                    + "The size of next ShuffleDescriptors: [{}].",
```
`ShuffleDescriptors` -> `blob`

Thesharing replied:
Resolved.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/BlobCacheSizeTracker.java:
```diff
@@ -99,44 +101,56 @@ public BlobCacheSizeTracker(long sizeLimit) {
+                total += size;
+                if (total > sizeLimit) {
+                    LOG.warn(
+                            "The overall size of ShuffleDescriptors in PermanentBlobCache exceeds "
```
`ShuffleDescriptors` -> `blob`

Thesharing replied:
Thank you for pointing this out. Resolved.

zhuzhurk commented on flink-runtime/src/main/java/org/apache/flink/runtime/blob/PermanentBlobCache.java:
```diff
@@ -247,6 +404,7 @@ public void run() {
 
                         boolean success = false;
                         try {
+                            blobCacheSizeTracker.unregisterJob(jobId);
```
Thanks for the explanation. I think you are right.

Thesharing reviewed (commented):
Thank you so much for re-reviewing my pull request. I've came up with several improvements according to your suggestions.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        // The size limit is the size of 2 blobs
+        int sizeLimit = 2;
```
This is not a size limit but a blob count limit, maybe `maxAcceptedBlobCount`?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            final Configuration blobClientConfig,
+            final BlobView blobView,
+            @Nullable final InetSocketAddress serverAddress,
+            final long sizeListOfBlobCacheSizeTracker)
```
Do you mean `sizeLimitOfBlobCacheSizeTracker `?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[4];
+            byte[][] data = new byte[4][BLOB_SIZE];
```
`BLOB_SIZE` does not make sense here. `data[i]` will the reassigned later.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            server.start();
+
+            // Initialize the data structures
+            JobID[] jobId = new JobID[4];
```
Let's create a variable to host this number 4.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            for (int i = 0; i < 4; i++) {
+                jobId[i] = new JobID();
+                data[i] = Arrays.copyOf(blobData, blobData.length);
+                data[i][i] ^= 1;
```
Maybe add a comments to say that this is to make contents different from each other for later verification?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            assertFalse(file[0].exists());
+            assertTrue(file[1].exists());
+
+            // Get the second blob once again, make the third blob to be the last recently used
```
`last` -> `least`

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            readFileAndVerifyContent(cache, jobId[1], key[1], data[1]);
+
+            // Then get the fourth blob, make sure the third blob is replaced.
+            // Here we cannot get the first blob again, because for transient blob,
```
The blobs are not transient here. IIUC, this file deletion is not related to BlobServer, right?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        final int concurrency = 3;
+        final ExecutorService executor = Executors.newFixedThreadPool(concurrency);
```
This kind of executor needs to be properly shutdown after used.

Thesharing replied:
Thank you for pointing this out. I should've notice issues like this. This test has been refactored.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+                jobId[i] = new JobID();
+
+                key[i] = put(server, jobId[i], blobData, blobType);
+                assertNotNull(key[i]);
```
all `blobData` are the same. Why are we not doing the same as in `testTrackSizeLimitAndDeleteExcess()`?

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,251 @@
+            byte[] blobData = new byte[BLOB_SIZE];
+            RANDOM.nextBytes(blobData);
+
+            for (int i = 0; i < 4; i++) {
```
Can we extract the blob initialization into a method to be reused by all tests?

Thesharing commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/BlobCacheSizeTrackerTest.java:
```diff
@@ -0,0 +1,168 @@
+    }
+
+    @Test
+    public void testZero() {
```
I feel so awful for these low-level issues. I could do better but I didn't. I should've examined and got them fixed in the early stage. I'll keep in mind and try to do better next time. Thank you so much for being patient and helping me improve the skill. BTW, this unit test has been improved.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,264 @@
+                BlobServerOptions.STORAGE_DIRECTORY, temporaryFolder.newFolder().getAbsolutePath());
+
+        try (BlobServer server = new BlobServer(config, new VoidBlobStore());
+                BlobCacheService cache =
```
Looks to me the `config` can be a class wide field. 
And seems `initBlobCacheServiceWithSizeLimit()` can be param-less.

zhuzhurk commented on flink-runtime/src/test/java/org/apache/flink/runtime/blob/PermanentBlobCacheSizeLimitTest.java:
```diff
@@ -0,0 +1,254 @@
+    private static final int MAX_NUM_OF_ACCEPTED_BLOBS = 2;
+    private static final int TOTAL_NUM_OF_BLOBS = 3;
+
+    private static final Configuration CONFIG = new Configuration();
```
I think it's better not to make `CONFIG` and `TEMPORARY_FOLDER` static and shared between tests.
It's safer to initialize them before each test.

Thesharing replied:
Yes, although the possibility of having the same JobID and BlobKey is low, but it still makes this case unstable. Resolved.

zhuzhurk commented:
@flinkbot run azure

zhuzhurk reviewed (approved):
Thanks for addressing all the comments @Thesharing 
The change now looks good to me. Will merge it once CI gives green.

Thesharing commented:
> Thanks for addressing all the comments @Thesharing
> The change now looks good to me. Will merge it once CI gives green.

Thank you for your reviewing, @zhuzhurk. It helps me a lot. The CI seems to be passed.
